copier/cli.py

```diff
@@ -48,57 +48,60 @@
 ```
 """
 
+import functools
 import sys
 from os import PathLike
 from pathlib import Path
 from textwrap import dedent
+from typing import Callable
 
 import yaml
-from decorator import decorator
 from plumbum import cli, colors
+from typing_extensions import ParamSpec
 
 from .errors import UnsafeTemplateError, UserMessageError
 from .main import Worker
 from .tools import copier_version
 from .types import AnyByStrDict, OptStr, StrSeq
 
+P = ParamSpec("P")
 
-@decorator
-def handle_exceptions(method, *args, **kwargs):
-    """Handle keyboard interruption while running a method."""
-    try:
+
+def _handle_exceptions(method: Callable[P, int]) -> Callable[P, int]:
+    @functools.wraps(method)
+    def inner(*args: P.args, **kwargs: P.kwargs) -> int:
         try:
-            return method(*args, **kwargs)
-        except KeyboardInterrupt:
-            raise UserMessageError("Execution stopped by user")
-    except UserMessageError as error:
-        print(colors.red | "\n".join(error.args), file=sys.stderr)
-        return 1
-    except UnsafeTemplateError as error:
-        print(colors.red | "\n".join(error.args), file=sys.stderr)
-        # DOCS https://github.com/copier-org/copier/issues/1328#issuecomment-1723214165
-        return 0b100
+            try:
+                return method(*args, **kwargs)
+            except KeyboardInterrupt:
+                raise UserMessageError("Execution stopped by user")
+        except UserMessageError as error:
+            print(colors.red | "\n".join(error.args), file=sys.stderr)
+            return 1
+        except UnsafeTemplateError as error:
+            print(colors.red | "\n".join(error.args), file=sys.stderr)
+            # DOCS https://github.com/copier-org/copier/issues/1328#issuecomment-1723214165
+            return 0b100
+
+    return inner
 
 
 class CopierApp(cli.Application):
     """The Copier CLI application."""
 
     DESCRIPTION = "Create a new project from a template."
-    DESCRIPTION_MORE = (
-        dedent(
-            """\
+    DESCRIPTION_MORE = dedent(
+        """\
             Docs in https://copier.readthedocs.io/
 
             """
-        )
-        + (
-            colors.yellow
-            | dedent(
-                """\
+    ) + (
+        colors.yellow
+        | dedent(
+            """\
```

Comment:
I don't think there should be any formatting changes here in this PR. Is perhaps your IDE using a more recent version of Ruff than Copier?

Reply:
that's likely the case. I wouldn't normally use lint-on-save anyway, but the `.vscode/settings.json` in copier's repo is intrusive.

configuring the repo to use an extension for formatting this way (instead of, say, `poetry run ruff format .`) effectively means you're using an unpinned formatter version and exposing PRs to these kinds of changes.

Reply:
Could you please revert those lines?

We should revise the VS Code settings in a separate PR. Perhaps there's a way to disable using the version of Ruff bundled with its VS Code extension.

Reply:
see https://github.com/copier-org/copier/pull/1514

turns out the ruff extension prefers the local ruff version *if present* and falls back to the bundled version. Since Ruff wasn't in the dev dependencies different contributors could get different versions depending on their setup